mtsmorf/io/move/read.py

```diff
@@ -0,0 +1,561 @@
+import json
+from dataclasses import dataclass
+from textwrap import shorten
+
+import mne
+import numpy as np
+import pandas as pd
+from mne import create_info
+from mne.io import RawArray
+from mne_bids import BIDSPath, read_raw_bids
+from mne_bids.read import _handle_events_reading
+from mne_bids.tsv_handler import _from_tsv
+
+from mtsmorf.io.move.utils import _preprocess_raw, _preprocess_epochs, _pl, _get_bad_epochs
+
+
+@dataclass()
+class Trial:
+    """Trial dataclass for the "move" efri task."""
+
+    trial_idx: int  # index in the behav.tsv
+    subject: str
+    perturbed: bool
+    fixation_trial: bool
+    target_direction: bool
+    success: bool
+
+    reaction_time: float  # (in seconds; time between go cue and left center)
+    correct_speed: bool  # (hard-flag on correct time)
+    speed_instruction: float
+    beh_fname: str
+    event_fname: str
+
+    # marks event markers of when this "trial"
+    # starts and ends
+    start_trial_id: str = None
+    end_trial_id: str = None
+    trial_duration: float = None
+    avg_speed: float = None
+    std_speed: float = None
+
+    def __repr__(self):
+        """Taken from mne-python."""
+        MAX_WIDTH = 68
+        strs = [f"<{self.subject} Move Trial {self.trial_idx} | %s non-empty values"]
+        non_empty = 0
+        for k, v in self.__dict__.items():
+            if k == "ch_names":
+                if v:
+                    entr = shorten(", ".join(v), MAX_WIDTH, placeholder=" ...")
+                else:
+                    entr = "[]"  # always show
+                    non_empty -= 1  # don't count as non-empty
+            elif k in ['index', 'subject']:
+                continue
+            elif k in ["speed_instruction"]:
+                entr = v
+            else:
+                try:
+                    this_len = len(v)
+                except TypeError:
+                    entr = "{}".format(v) if v is not None else ""
+                else:
+                    if this_len > 0:
+                        entr = "%d item%s (%s)" % (
+                            this_len,
+                            _pl(this_len),
+                            type(v).__name__,
+                        )
+                    else:
+                        entr = ""
+            if entr != "":
+                non_empty += 1
+                strs.append("%s: %s" % (k, entr))
+        st = "\n ".join(sorted(strs))
+        st += "\n>"
+        st %= non_empty
+        return st
+
+    def to_data_frame(self):
+        df = pd.DataFrame.from_dict(self.__dict__.items(), orient='columns').T
+        df.columns = df.iloc[0]
+
+        # set datatype
+        df['beh_fname'] = df['beh_fname'].astype(str)
+        df['event_fname'] = df['event_fname'].astype(str)
+
+        df.drop(df.index[0], inplace=True)
+        return df
+
+    def _add_xy_metrics(self, metrics_df):
+        if not self.success or self.perturbed:
+            return
```

Comment:
would return a warning instead of failing silently:

`warnings.warn("xy metrics not added because Trial instance includes unsuccessful or perturbed trials.)`

just `import warnings` if you include this